### ISSUE TITLE
Connections from binaries

### ISSUE THREAD

**Issue by anpic** [celzero/rethink-app#1163]
Are you planning to add connection blocking for system binaries for example such as ping or curl?
It works in firewalls like that PCAPdroid and NoRoot Firewall. There is a block even if you run binaries under the root.

**Comment by ignoramous** [comment]
I don't understand what you mean by "binaries"? Are they installed as apps in Android? Or installed from within tmux or side-loaded to `/usr/bin` etc on a rooted device?

**Comment by anpic** [comment]
> I don't understand what you mean by "binaries"?

For example /system/bin/ping

**Comment by ignoramous** [comment]
unless binaries have their own UID (user id), Rethink can't do much. But otherwise, all *new* UIDs that Rethink "sees" will appear in "Apps" list. 

You can enable *Block newly installed apps by default* to make sure these *new* UIDs get blocked unless explicitly allowed.

**Comment by anpic** [comment]
> unless binaries have their own UID (user id), Rethink can't do much

But PCAPdroid with exactly the same capabilities does it ;)

**Comment by ignoramous** [comment]
> But PCAPdroid with exactly the same capabilities does it

I don't use pcapdroid or familiar with its code, so I can't say just what it is doing. If pcapdroid developer has any information on this, I'll gladly review and see if we can implement it pronto.

> This can be done by firewalls that generally do without permission to access the Internet

I don't understand how this is related to what we are discussing? Rethink is not really a firewall in the traditional sense. Technically, one can say it is but a *Network Monitor* with Firewall-like capabilities.

**Comment by anpic** [comment]
> I'll gladly review and see if we can implement it pronto

https://emanuele-f.github.io/PCAPdroid/

> Technically, one can say it is but a _Network Monitor_ with Firewall-like capabilities.

Oh, that's how it's :)

>DNS over HTTPS / DNS over Tor / DNSCrypt client, WireGuard proxifier, firewall, and connection tracker for Android. 
>Rethink DNS + Firewall + VPN for Android

Then you need to fix everything: Firewall --> Firewall-like capabilities ;)

**Comment by ignoramous** [comment]
>  Then you need to fix everything: Firewall --> Firewall-like capabilities 

Do you know what a TUN/TAP device is? If so, you'd understand why Rethink is firewall. Do you know what a packet filter is? If so, you'd understand why Rethink relatively only has firewall-like capabilities. If not, I can see why what I'm saying can be confusing.

> I'll gladly review and see if we can implement it pronto

> > https://emanuele-f.github.io/PCAPdroid/

Let me know once Emanuele tells you how binaries are being blocked, and points you to the code where pcapdroid identifies a "binary", because I don't understand how pcapdroid does it (but clearly, it is possible).

**Comment by anpic** [comment]
> If not, I can see why what I'm saying can be confusing.

Confusing reality :)
Can you show me how to block all traffic? Emanuele can.

>If so, you'd understand why Rethink is firewall
I perfectly understand the difference between a firewall and PR.
Emanuele doesn't do much PR but he has already done a lot alone.
You have a lot of PR but when it comes to checking the main function then I see traffic leaks. And even in DNS traffic. And I'm trying to figure out the reasons and settings.
But instead I hear in response that you no longer have a firewall at all but "Firewall-like capabilities".
And I'm starting to doubt everything.
And here it's your choice: either try to dispel my doubts or pretend that these are all my problems and I have to figure everything out on my own.

> Let me know once Emanuele tells you

Do I understand correctly that I have to discuss everything with Emanuele for you and then bring everything to you ready-made?
In fact, a lot is written in PCAPdroid documentation itself but you should at least read it ;)

**Comment by anpic** [comment]
> If not, I can see why what I'm saying can be confusing.

Confusing reality :)
Can you show me how to block all traffic? Emanuele can.

>If so, you'd understand why Rethink is firewall

I perfectly understand the difference between a firewall and PR.
Emanuele doesn't do much PR but he has already done a lot alone.
You have a lot of PR but when it comes to checking the main function then I see traffic leaks. And even in DNS traffic. And I'm trying to figure out the reasons and settings.
But instead I hear in response that you no longer have a firewall at all but "Firewall-like capabilities".
And I'm starting to doubt everything.
And here it's your choice: either try to dispel my doubts or pretend that these are all my problems and I have to figure everything out on my own.
But I think all users would be grateful to you for detailed docs on how your product can block maximum traffic.

> Let me know once Emanuele tells you

Do I understand correctly that I have to discuss everything with @emanuele-f for you and then bring everything to you ready-made?
In fact, a lot is written in PCAPdroid documentation itself but you should at least read it ;)

**Comment by ignoramous** [comment]
Some background: I worked on Android OS for 5 years. I'm very familiar with the internals.

> Can you show me how to block all traffic? Emanuele can.

Emanuele is ignorant of how Android works. There's no app that can block all traffic.

> You have a lot of PR but when it comes to checking the main function then I see traffic leaks

Please point to those traffic leaks. We'll fix them.

> And I'm trying to figure out the reasons and settings.

*GitHub Issues* is not the place. Prefer  asking our [subreddit](https://reddit.com/r/rethinkdns) or [telegram community](https://t.me/rethinkdns) for Q&A.

> But I think all users would be grateful to you for detailed docs on how your product can block maximum traffic.

My bad. I thought people are merely grateful for the fact that we put up the code and gave away the app for free, but what do I know.

> Do I understand correctly that I have to discuss everything with @emanuele-f for you and then bring everything to you ready-made

No. You misunderstand. What I said is, I do not use pcapdroid, so I don't know what you're saying is even correct. If it is, clarify it with Emanuele and then get back with code references, if possible. My assessment remains that *binaries* cannot be blocked by firewall using the VPN APIs unless those binaries are "installed" as *apps* (which they aren't).

> In fact, a lot is written in PCAPdroid documentation itself but you should at least read it

Like I said, if you have, point me to the relevant bits.

**Comment by emanuele-f** [comment]
Guys, idk what problem you have, but it's so weird for me to be called in cause and insulted. @ignoramous does it look normal to you? Please stop tagging me in and be respectful, I don't know you nor I ever judged your knowlege, work or experience

**Comment by emanuele-f** [comment]
@anpic just my suggestion, you may not realize, but your comments are a bit too "strong". It's good to express your opinion, but you should understand on the other side there are people with their opinion and your comments may seem like an attack

**Comment by anpic** [comment]
> Please stop tagging me in and be respectful

I'm sorry, it's my fault, it won't happen again.

> but you should understand on the other side there are people with their opinion and your comments may seem like an attack

I was just expressing my opinion. Respectfully but without ambiguity and sincerely. Just like when I was talking to you.

**Comment by anpic** [comment]
> Some background: I worked on Android OS for 5 years. I'm very familiar with the internals.

I didn't say anything about your competencies.
But there's a difference between knowing the firewall and creating the firewall.

> Emanuele is ignorant of how Android works.

Ah, that's it..
Can you tell us more about the perfection of Android API in general and VpnService in particular?

> There's no app that can block all traffic.

Really? If you say this for sure then can I see your comparative testing in detail?
As an example of testing high-level specialists.

**Comment by anpic** [comment]
> Please point to those traffic leaks. We'll fix them.

You told me yourself above that you don't have a firewall but only its firewall-like capabilities.
How are you going to stop leaks from binaries?
How do I set up at least a complete DNS connection block?

> _GitHub Issues_ is not the place.

Why should I go somewhere else if there is a universal platform for any discussions on development and their configuration?
Why not post wiki and docs right here like other developers do?

> My bad. I thought people are merely grateful for the fact that we put up the code and gave away the app for free, but what do I know.

This is a fair statement.
But I'm sure that you expressed it here for the first time.
And all in order to get away from a detailed technical discussion of your product?

**Comment by emanuele-f** [comment]
@anpic you are not considering all the hard work that was done to bring this project to live. There is no duty from the developer to reply to your queries, to support people, to continue to develop, most of this is done for free, not considering the risk of other people stealing your code without even mentioning you.
If you realize all of this, it will be clear that you cannot approach people like this and you will respect all the other open source developers. If you think you can do better, you have the source. Develop it.
I am unsubscribing from this thread. After all of these crititcs, it's almost impossible that this thread can bring any value to this project.

**Comment by anpic** [comment]
> My assessment remains that _binaries_ cannot be blocked by firewall using the VPN APIs unless those binaries are "installed" as _apps_ (which they aren't).

Yes, they can't, say thank you to your favorite company.
Therefore he had to apply competencies that few specialists seem to have...

> Like I said, if you have, point me to the relevant bits.

Still it's interesting: you are a top-level specialist but I have to tell you the details :)
I'm willing to take the time to discuss the details.
But I don't want to waste time and end up seeing this discussion closed and deleted. As specialists sometimes do who turn out to be not really specialists..
Will you promise that this discussion will remain public?

**Comment by anpic** [comment]
> you are not considering all the hard work that was done to bring this project to live. There is no duty from the developer to reply to your queries, to support people, to continue to develop, most of this is done for free, not considering the risk of other people stealing your code without even mentioning you.

But for some reason, some developers create docs and answer inconvenient questions while others dodge them and pretend that they are doing a generous thing on their own.
And if the case turns out to be not as generous as they are trying to imagine?

**Comment by ignoramous** [comment]
> Guys, idk what problem you have, but it's so weird for me to be called in cause and insulted. @ignoramous does it look normal to you? 

No one has a problem here. There isn't one insult hurled at you. If my comment that someone is *'ignorant of how Android works' if they claim a non-root app can block ALL traffic* is insulting, I don't know what to say.

> I was just expressing my opinion. Respectfully but without ambiguity and sincerely. Just like when I was talking to you.

No, you are not.

> But there's a difference between knowing the firewall and creating the firewall.

Sure.

> How are you going to stop leaks from binaries?

I do not know how this is even possible. Can you share a screen recording of any app that does this.

> How do I set up at least a complete DNS connection block?

You can't "completely" block anything with Rethink. There are many reasons for this. For the traffic that is sent to Rethink by Android, you can test if adding a `*` in universal domain rules works. I haven't tried it. If it doesn't work, then we should implement it.

> Why should I go somewhere else if there is a universal platform for any discussions on development and their configuration?

This is Github "issues".

> Why not post wiki and docs right here like other developers do?

The app isn't `v1` yet and keeps adding newer features, changing /correcting its behaviour, release over release, making maintaining docs hard. Perhaps, when we are `v1` and UI has settled into something better, we would put docs up.

> And all in order to get away from a detailed technical discussion of your product?

For a technical discussion, I need to be talking to folks who are technical enough. From your responses, it doesn't look like you are; but I may be wrong.

> Still it's interesting: you are a top-level specialist but I have to tell you the details :)

Because you use the app, and claim it has this functionality. I am merely asking for pointers. If you have none, just say you have none; and we'd be done. There's no obligation on your part.

> Your seniority and coins are not important here, factual knowledge is important here.

Sure.

**Comment by anpic** [comment]
> For a technical discussion, I need to be talking to folks who are technical enough. From your responses, it doesn't look like you are; but I may be wrong.

Okay, let's try it ;)

> Can you share a screen recording of any app that does this.

https://user-images.githubusercontent.com/5488003/233846260-64ffd9c7-fbb7-4c25-93df-e463c0b18f76.png
You know what NETD is, right?
And there are also items with the word Unknown ;)

> You can't "completely" block anything with Rethink

You should have said that right away :)
But then it would be nice to stop promoting with the word Firewall ;)
It looks like you weren't originally going to make a firewall. But how else to lure users?

> Because you use the app, and claim it has this functionality. I am merely asking for pointers.

It seemed to me easy to take and read the documentation for PCAPdroid:

> When running in non-root mode, PCAPdroid alters the network traffic, which is a required step in order to capture it via the VPNService. The modifications only affect the L3 and L4 layers of the packets, whereas the L7 data is untouched.

> For a high-level overview

The question was about something completely different.
Do you understand the problems specifically created at the Android API level?
Do you understand the limitations that your favorite company creates for those who are trying to make a full-fledged firewall?

> I know from the workings of the AOSP code. For some (non-exhaustive) pointers on this, see #224

This is a good issue.
And if you wanted to create a firewall then you would undoubtedly get acquainted with the masterpiece that was created in NetGuard. The worthy heir of which was PCAPdroid.
Or will you also say that @M66B

> is ignorant of how Android works

?
Marcel Bokhorst is one of the greatest android developers of our time. And he was the first to prove that tracking all outgoing traffic is possible even in a noroot mode.

**Comment by ignoramous** [comment]
>  You know what NETD is, right?

These are part of some 7 hardcoded names in pcapdroid ([ref](https://github.com/emanuele-f/PCAPdroid/blob/40dd43d2148bde0fe1a1f235e453a422c1bc4ad0/app/src/main/java/com/emanuelef/remote_capture/AppsResolver.java#L68-L90)). Rethink uses names for some 100+ such components (not just 7) as defined in AOSP ([ref](https://github.com/celzero/rethink-app/blob/859519a7b051c185c42b8c6bad46e8a5d7ef395c/app/src/main/java/com/celzero/bravedns/util/AndroidUidConfig.kt#L18-L148). These are shown *only* when any of these components try to establish a connection. These are not "binaries".

> And there are also items with the word Unknown ;)

You'd see *Unknown* frequently with Rethink; these aren't binaries. This is a result of the Android OS losing account of the real owner of a given outgoing connection. This is one reason why Rethink has a setting to *Block when source is unknown* setting.

>  But then it would be nice to stop promoting with the word Firewall ;)

Like I said, there are many types of firewall. Rethink is specifically a userspace stateful Firewall (just like NetGuard and TrackerControl). The kind of firewall you're thinking about (what I call a 'traditional firewall' in the readme linked to above) isn't possible on Android without `root` (use [AfWall+](https://github.com/ukanth/afwall)).

> But of course, in order to do this, he had to surpass even Marcel in these competencies.

Sure.

**Comment by anpic** [comment]
> These are part of some 7 hardcoded names in pcapdroid ([ref](https://github.com/emanuele-f/PCAPdroid/blob/40dd43d2148bde0fe1a1f235e453a422c1bc4ad0/app/src/main/java/com/emanuelef/remote_capture/AppsResolver.java#L68-L90)). Rethink uses names for some 100+ such components (not just 7) as defined in AOSP ([ref](https://github.com/celzero/rethink-app/blob/859519a7b051c185c42b8c6bad46e8a5d7ef395c/app/src/main/java/com/celzero/bravedns/util/AndroidUidConfig.kt#L18-L148)). These are shown _only_ when any of these components try to establish a connection.

But let's look at this point first.
How do you filter applications by all these uids?
After all, you said yourself above that you respect Android API.
But standard VpnService doesn't allow to filter by uid.

**Comment by ignoramous** [comment]
>  How do you filter applications by all these uids?

These don't show up on our app unless there's at least one outgoing connection. If you enable `Block newly installed apps by default* then these uid would get blocked whenever they show up. We'll consider showing "some" of these UIDs upfront: #1167 

>  But standard VpnService doesn't allow to filter by uid.

You're hearing things no one has said.

> But this discussion here has long gone beyond binaries :)

Please a different discussion (or open an issue, if there's one). I'm closing this one in favour of #1167.
